brokers/vscode/broker.go

```diff
@@ -0,0 +1,296 @@
+	if err != nil {
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+	return broker.injectRemotePlugin(meta, unpackedPath, image, pj)
+}
+
+func (broker *VSCodeExtensionBroker) getPackageJSON(pluginFolder string) (*packageJSON, error) {
+	packageJSONPath := filepath.Join(pluginFolder, "extension", "package.json")
+	broker.PrintDebug("Reading package.json of VS Code extension from '%s'", packageJSONPath)
+	f, err := ioutil.ReadFile(packageJSONPath)
+	if err != nil {
+		return nil, err
+	}
+	pj := &packageJSON{}
+	err = json.Unmarshal(f, pj)
+	return pj, err
+}
+
+func (broker *VSCodeExtensionBroker) injectRemotePlugin(meta model.PluginMeta, archiveFolder string, image string, pj *packageJSON) error {
+	pluginFolderPath := filepath.Join("/plugins", fmt.Sprintf("%s.%s", meta.ID, meta.Version))
+	broker.PrintDebug("Copying VS Code extension '%s:%s' from '%s' to '%s'", meta.ID, meta.Version, archiveFolder, pluginFolderPath)
+	err := broker.ioUtil.CopyResource(archiveFolder, pluginFolderPath)
+	if err != nil {
+		return err
+	}
+	tooling := &model.ToolingConf{
+		Containers: []model.Container{*containerConfig(image)},
+	}
+	broker.addPortToTooling(tooling, pj)
+	return broker.Storage.AddPlugin(&meta, tooling)
+}
+
+func containerConfig(image string) *model.Container {
+	c := model.Container{
+		Name:  "vscodeextsidecar" + randomNumberAsString(),
+		Image: image,
+		Volumes: []model.Volume{
+			{
+				Name:      "projects",
+				MountPath: "/projects",
+			},
+			{
+				Name:      "plugins",
+				MountPath: "/plugins",
+			},
+		},
+	}
+	return &c
+}
+
+func (broker *VSCodeExtensionBroker) addPortToTooling(toolingConf *model.ToolingConf, pj *packageJSON) {
+	port := findPort()
+	sPort := strconv.Itoa(port)
+	endpointName := "port" + sPort
+	var re = regexp.MustCompile(`[^a-z_0-9]+`)
+	prettyID := re.ReplaceAllString(pj.Publisher+"_"+pj.Name, `_`)
+	theiaEnvVar1 := "THEIA_PLUGIN_REMOTE_ENDPOINT_" + prettyID
+	theiaEnvVarValue := "ws://" + endpointName + ":" + sPort
+
+	toolingConf.Containers[0].Ports = append(toolingConf.Containers[0].Ports, model.ExposedPort{ExposedPort: port})
+	toolingConf.Endpoints = append(toolingConf.Endpoints, model.Endpoint{
+		Name:       endpointName,
+		Public:     false,
+		TargetPort: port,
+	})
+	toolingConf.Containers[0].Env = append(toolingConf.Containers[0].Env, model.EnvVar{Name: "THEIA_PLUGIN_ENDPOINT_PORT", Value: sPort})
+	toolingConf.WorkspaceEnv = append(toolingConf.WorkspaceEnv, model.EnvVar{Name: theiaEnvVar1, Value: theiaEnvVarValue})
+}
+
+func randomNumberAsString() string {
+	port := findPort()
+	return strconv.Itoa(port)
+}
+
+func findPort() int {
+	return 4000 + rand.Intn(6000)
+}
+
+func (broker *VSCodeExtensionBroker) download(extension string, dest string, meta model.PluginMeta) error {
+	response, err := fetchExtensionInfo(extension, meta)
+	if err != nil {
+		return err
+	}
+
+	URL, err := findAssetURL(response, meta)
+	if err != nil {
+		return err
```

Comment:
If something went wrong it will just propagate an exception `bad json` what is not very helpful message. Please consider adding context message here. I mean `Error failed during searching asset URL. Cause: ...`